tests/subsys/fs/nvs/src/main.c

```diff
@@ -959,3 +959,213 @@ ZTEST_F(nvs, test_nvs_cache_hash_quality)
+	zassert_equal(ret, sizeof(item0_data), "nvs_write() failed: %d", ret);
+
+	/*
+	 * Write an item of which ATE has a valid CRC but contains corrupted data (to simulate a
+	 * 8-bit CRC issue)
+	 */
+	ate.id = 1;
+	callback_corrupt_ate(&ate, items); /* Corrupt the ATE offset and/or length */
+	ate.crc8 = crc8_ccitt(0xff, &ate, offsetof(struct nvs_ate, crc8));
+
+	/* Write the data part */
+	data_crc = crc32_ieee((const uint8_t *) item1_data, sizeof(item1_data));
+	memcpy(buf, item1_data, sizeof(item1_data));
+	memcpy(buf + sizeof(item1_data), &data_crc, sizeof(data_crc));
+	ret = flash_write(fixture->fs.flash_device, fixture->fs.offset + sizeof(item0_data) +
+			  sizeof(uint32_t), buf, sizeof(item1_data) + sizeof(data_crc));
+	zassert_true(ret == 0,  "flash_write() failed: %d", ret);
+
+	/* Write the ATE part */
+	ret = flash_write(fixture->fs.flash_device, fixture->fs.offset + fixture->fs.sector_size -
+			  sizeof(struct nvs_ate) * 4, &ate, sizeof(ate));
+	zassert_true(ret == 0,  "flash_write() failed: %d", ret);
+
+	/* Update the NVS */
+	fixture->fs.data_wra += sizeof(item1_data) + sizeof(data_crc);
+	fixture->fs.ate_wra -= sizeof(struct nvs_ate);
+
+	/* Write two more good items */
+	ret = nvs_write(&fixture->fs, 2, item2_data, sizeof(item2_data));
+	zassert_equal(ret, sizeof(item2_data), "nvs_write() failed: %d", ret);
+	ret = nvs_write(&fixture->fs, 3, item3_data, sizeof(item3_data));
+	zassert_equal(ret, sizeof(item3_data), "nvs_write() failed: %d", ret);
+
+	for (i = 0; i < ARRAY_SIZE(items); i++) {
+		memset(buf, 0, sizeof(buf));
+		ret = nvs_read(&fixture->fs, i, buf, sizeof(buf));
+		if (i == 1) {
+			zassert_equal(ret, -EIO,
+				      "nvs_read() failed with %d for bad item %d", ret, i);
+		} else {
+			zassert_true(ret >= 0,
+				     "nvs_read() failed with %d for good item %d", ret, i);
+			zassert_true(memcmp(items[i].data, buf, items[i].size) == 0,
+				     "item %d data is corrupted", i);
+		}
+	}
+
+	/* Mount does not detect the issue */
+	ret = nvs_mount(&fixture->fs);
+	zassert_true(ret == 0, "nvs_mount() failed: %d", ret);
+
+	/* Fill what remains of the sector (the GC is called as soon as a sector is closed because
+	 * there is no more room in it to write an item). Take into account the test data and the
+	 * mandatory reserved ATEs.
+	 * It is not possible to fill the sectors by just calling nvs_write() and expecting it to
+	 * return ENOSPC, because the GC does not terminate due to the corrupted length
+	 */
+	zassert_true(fixture->fs.ate_wra >> ADDR_SECT_SHIFT == 0, "bad current sector %d",
+		     fixture->fs.ate_wra >> ADDR_SECT_SHIFT);
+	data_bytes_per_sector = fixture->fs.sector_size - (4 * sizeof(struct nvs_ate));
+	total_free_data_bytes = data_bytes_per_sector * (fixture->fs.sector_count - 1) -
+				(sizeof(item0_data) + sizeof(item1_data) + sizeof(item2_data) +
+				sizeof(item3_data) + (4 * sizeof(struct nvs_ate)));
+	count = total_free_data_bytes / (sizeof(i) + sizeof(struct nvs_ate));
```

Comment:
It is confusing here to use index i
You are also going need to take into account CRC32, right ?